go/vt/sqlparser/ast_funcs.go

```diff
@@ -562,6 +563,20 @@ func NewTypedArgument(in string, t sqltypes.Type) *Argument {
 	return &Argument{Name: in, Type: t}
 }
 
+func NewTypedArgumentFromLiteral(in string, lit *Literal) (*Argument, error) {
+	arg := &Argument{Name: in, Type: lit.SQLType()}
+	switch arg.Type {
+	case sqltypes.Decimal:
+		dec, err := decimal.NewFromMySQL(lit.Bytes())
+		if err != nil {
+			return nil, err
+		}
+		arg.Scale = -dec.Exponent()
+		arg.Size = dec.Size()
+	}
+	return arg, nil
```

Comment:
Since this is going to be working off literals off a MySQL expression, I think they can be assumed to have a valid Decimal format. Hence, it'd be both significantly cheaper and simpler to calculate the Scale and Size by counting the digits in `lit.Bytes()` rather than parsing a whole Decimal just to extract its size + scale. I would add a new helper method to the `decimal` package that does just that.